log/loki.go

```diff
@@ -0,0 +1,499 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+type lokiHook struct {
+	addr       string
+	labels     [][2]string
+	ch         chan *logrus.Entry
+	limit      int
+	msgMaxSize int
+	levels     []logrus.Level
+	pushPeriod time.Duration
+	client     *http.Client
+	ctx        context.Context
+	profile    bool
+}
+
+// LokiFromConfigLine returns a new logrus.Hook that pushes logrus.Entrys to loki and is configured
+// through the provided line
+func LokiFromConfigLine(ctx context.Context, line string) (logrus.Hook, error) {
+	h := &lokiHook{
+		addr:       "http://127.0.0.1:3100/loki/api/v1/push",
+		limit:      100,
+		levels:     logrus.AllLevels,
+		pushPeriod: time.Second * 1,
+		ctx:        ctx,
+		msgMaxSize: 1024 * 1024, // 1mb
+	}
+	if line == "loki" {
+		return h, nil
+	}
+
+	parts := strings.SplitN(line, "=", 2)
+	if parts[0] != "loki" {
+		return nil, fmt.Errorf("loki configuration should be in the form `loki=url-to-push` but is `%s`", line)
+	}
+	args := strings.Split(parts[1], ",")
+	h.addr = args[0]
+	// TODO use something better ... maybe
+	// https://godoc.org/github.com/kubernetes/helm/pkg/strvals
+	// atleast until https://github.com/loadimpact/k6/issues/926?
+	if len(args) == 1 {
+		return h, nil
+	}
+
+	for _, arg := range args[1:] {
+		paramParts := strings.SplitN(arg, "=", 2)
+
+		if len(paramParts) != 2 {
+			return nil, fmt.Errorf("loki arguments should be in the form `address,key1=value1,key2=value2`, got %s", arg)
+		}
+
+		key := paramParts[0]
+		value := paramParts[1]
+
+		var err error
+		switch key {
+		case "pushPeriod":
+			h.pushPeriod, err = time.ParseDuration(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki pushPeriod %w", err)
+			}
+		case "profile":
+			h.profile = true
+		case "limit":
+			h.limit, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki limit as a number %w", err)
+			}
+		case "msgMaxSize":
+			h.msgMaxSize, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki msgMaxSize as a number %w", err)
+			}
+		case "level":
+			h.levels, err = getLevels(value)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			if strings.HasPrefix(key, "label.") {
+				labelKey := strings.TrimPrefix(key, "label.")
+				h.labels = append(h.labels, [2]string{labelKey, value})
+				continue
+			}
+
+			return nil, fmt.Errorf("unknown loki config key %s", key)
+		}
+	}
+
+	h.client = &http.Client{Timeout: h.pushPeriod}
+
+	h.start()
+	return h, nil
+}
+
+func getLevels(level string) ([]logrus.Level, error) {
+	// TODO figure out if `tracing`,`fatal` and `panic` should be included
```

Comment:
Is there a reason _not_ to include them, besides adding more cases to the switch and making it even uglier :stuck_out_tongue_winking_eye: I don't think we currently use `log.Fatal()` anywhere, though if I didn't want to return a custom exit code, I imagine I'd probably have used it in here: https://github.com/loadimpact/k6/blob/8f4797d2761560fefb3e01c1bf440ee6a1c55548/cmd/run.go#L261-L263

In any case, it seems prudent to not omit `fatal` here, since if we ever start using it in the future, I imagine it's going to be the most important message our users will want to see from the whole log :sweat_smile: 

Even more importantly, we currently use `panic()` a lot in k6, mostly in places which supposedly should never be reached. I didn't realize logrus had `Panic*()` methods, but now that I do, part of the logging refactoring should be to switch to `logger.Panic()` where we can. This way, and if you add panic level here, if we ever introduce a bug in the future that causes k6 to hit one of these panics, there's a good chance our users will see it. Due to the log buffering, we might have to implement some flushing in a `defer / recover()` somewhere, but it's mostly doable... 

So, yes, in summary, please include at least `fatal` and `panic` levels here - we probably are never going to care about tracing, unless we want to try and make k6 less noisy and hide some internal debug messages.

Reply:
neither one of those is likely to work ATM as the process will end before we push it to loki ...

Reply:
I realize,
> Due to the log buffering, we might have to implement some flushing in a defer / recover() somewhere, but it's mostly doable...

I still don't think it's a good reason to arbitrarily omit log levels here, now :man_shrugging:

Reply:
Didn't realize `ParseLevel()` existed, :heart: this approach, and sorry for badgering you into removing the switch